### PR TITLE
KDL IK solver cleanup

### PR DESCRIPTION
**Pull request by rhaschke**
Tracking down #1255 and in preparation of #1208, I first cleaned up the KDL IK solver code
- removing dead code
- removing redundant joint handling
- merging ChainIkSolverPos into main plugin source
- use standard Eigen::SVD that yields higher accuracy

These changes should not affect functionality as verified by kinematics unit tests (#1272). The cleanup yields a strong performance boost (running 10.000 deterministic IK attempts):
- original code: 410ms
- after cleanup: 230ms
- standard Eigen::SVD: 240ms

**Please merge in conjuction with #1272 and #1288.**

### CODE REVIEW

File: moveit_kinematics/kdl_kinematics_plugin/include/moveit/kdl_kinematics_plugin/chainiksolver_vel_pinv_mimic.hpp [hunk 1]
```diff
@@ -51,21 +51,15 @@ class ChainIkSolverVel_pinv_mimic : public ChainIkSolverVel
...
+   * @param chain the chain to calculate the inverse velocity kinematics for
+   * @param num_mimic_joints The number of joints that are setup to follow other joints
+   * @param position_ik false if you want to solve for the full 6 dof end-effector pose,
+   *        true if you want to solve only for the 3 dof end-effector position.
+   * @param eps if a singular value is below this value, its inverse is set to zero, default: 0.00001
+   * @param maxiter maximum iterations for the svd calculation, default: 150
    */
-  explicit ChainIkSolverVel_pinv_mimic(const Chain& chain, int num_mimic_joints = 0, int num_redundant_joints = 0,
-                                       bool position_ik = false, double eps = 0.00001, int maxiter = 150);
+  explicit ChainIkSolverVel_pinv_mimic(const Chain& chain, int num_mimic_joints = 0, bool position_ik = false,
```
Comment (mlautman) [inline moveit_kinematics/kdl_kinematics_plugin/include/moveit/kdl_kinematics_plugin/chainiksolver_vel_pinv_mimic.hpp]: I know this isn't from this PR but why does this method name mix styles?

File: moveit_kinematics/kdl_kinematics_plugin/src/kdl_kinematics_plugin.cpp [hunk 2]
```diff
@@ -303,18 +297,25 @@ bool KDLKinematicsPlugin::searchPositionIK(const geometry_msgs::Pose& ik_pose, c
     error_code.val = error_code.NO_IK_SOLUTION;
     return false;
   }
+  // Resize consistency limits to remove mimic joints
+  std::vector<double> consistency_limits_mimic;
+  if (!consistency_limits.empty())
+    for (std::size_t i = 0; i < dimension_; ++i)
```
Comment (mlautman) [inline moveit_kinematics/kdl_kinematics_plugin/src/kdl_kinematics_plugin.cpp]: Add `{}` to the if statement

**Comment by mlautman** [comment]
This contribution was clearly a ton of work and makes a ton of improvements. That said, it is really hard to review this for correctness. Can you please add tests to demonstrate the correctness of these changes?

File: moveit_kinematics/kdl_kinematics_plugin/include/moveit/kdl_kinematics_plugin/chainiksolver_vel_pinv_mimic.hpp (hunk 1 above)
Comment (rhaschke) [reply moveit_kinematics/kdl_kinematics_plugin/include/moveit/kdl_kinematics_plugin/chainiksolver_vel_pinv_mimic.hpp]: I don't know either. Which name do you suggest?

**Comment by rhaschke** [comment]
> Can you please add tests to demonstrate the correctness of these changes?

As already written in the main comment: These changes are verified by kinematics unit tests (#1272).

File: moveit_kinematics/kdl_kinematics_plugin/include/moveit/kdl_kinematics_plugin/chainiksolver_vel_pinv_mimic.hpp (hunk 1 above)
Comment (rhaschke) [reply moveit_kinematics/kdl_kinematics_plugin/include/moveit/kdl_kinematics_plugin/chainiksolver_vel_pinv_mimic.hpp]: I renamed it to `ChainIkSolverVelMimicSVD` now.

File: moveit_kinematics/kdl_kinematics_plugin/src/kdl_kinematics_plugin.cpp (hunk 2 above)
Comment (rhaschke) [reply moveit_kinematics/kdl_kinematics_plugin/src/kdl_kinematics_plugin.cpp]: Added.

File: moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h [hunk 3]
```diff
@@ -149,19 +149,20 @@ class KinematicsBase
   static const double DEFAULT_SEARCH_DISCRETIZATION; /* = 0.1 */
   static const double DEFAULT_TIMEOUT;               /* = 1.0 */
 
-  /** @brief The signature for a callback that can compute IK */
+  /** @brief Signature for a callback to validate an IK solution */
```
Comment (davetcoleman) [inline moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h]:
```suggestion
  /** @brief Signature for a callback to validate an IK solution. Typically used for collision checking. */
```

File: moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h [hunk 4]
```diff
@@ -149,19 +149,20 @@ class KinematicsBase
...
    * @brief Given a desired pose of the end-effector, compute the joint angles to reach it
+   *
+   * In contrast to the searchPositionIK methods, this one is expected to return the solution
+   * closest to the seed state. Randomly re-seeding is explicitly not allowed.
    * @param ik_pose the desired pose of the link
    * @param ik_seed_state an initial guess solution for the inverse kinematics
    * @param solution the solution vector
    * @param error_code an error code that encodes the reason for failure or success
-   * @param lock_redundant_joints if setRedundantJoints() was previously called, keep the values of the joints marked as
-   * redundant the same as in the seed
```
Comment (davetcoleman) [inline moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h]:
```suggestion
   * @param options An option struct which contains the type of redundancy discretization used. This default
   *                implementation only supports the KinematicSearches::NO_DISCRETIZATION method; requesting any
   *                other will result in failure.
```

File: moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h [hunk 5]
```diff
@@ -190,15 +191,13 @@ class KinematicsBase
...
   /**
    * @brief Given a desired pose of the end-effector, search for the joint angles required to reach it.
-   * This particular method is intended for "searching" for a solutions by stepping through the redundancy
+   * This particular method is intended for "searching" for a solution by stepping through the redundancy
    * (or other numerical routines).
    * @param ik_pose the desired pose of the link
    * @param ik_seed_state an initial guess solution for the inverse kinematics
    * @param timeout The amount of time (in seconds) available to the solver
    * @param solution the solution vector
    * @param error_code an error code that encodes the reason for failure or success
```
Comment (davetcoleman) [inline moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h]:
```suggestion
    * @param error_code an error code that encodes the reason for failure or success
    * @param options An option struct which contains the type of redundancy discretization used. This default
    *                implementation only supports the KinematicSearches::NO_DISCRETIZATION method; requesting any
    *                other will result in failure.
```

File: moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h [hunk 6]
```diff
@@ -217,8 +216,6 @@ class KinematicsBase
    * current seed state
    * @param solution the solution vector
    * @param error_code an error code that encodes the reason for failure or success
```
Comment (davetcoleman) [inline moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h]:
```suggestion
    * @param error_code an error code that encodes the reason for failure or success
    * @param options An option struct which contains the type of redundancy discretization used. This default
    *                implementation only supports the KinematicSearches::NO_DISCRETIZATION method; requesting any
    *                other will result in failure.
```

File: moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h [hunk 7]
```diff
@@ -229,16 +226,14 @@ class KinematicsBase
...
-   * This particular method is intended for "searching" for a solutions by stepping through the redundancy
+   * This particular method is intended for "searching" for a solution by stepping through the redundancy
    * (or other numerical routines).
    * @param ik_pose the desired pose of the link
    * @param ik_seed_state an initial guess solution for the inverse kinematics
    * @param timeout The amount of time (in seconds) available to the solver
    * @param solution the solution vector
-   * @param solution_callback A callback solution for the IK solution
+   * @param solution_callback A callback to validate an IK solution
    * @param error_code an error code that encodes the reason for failure or success
```
Comment (davetcoleman) [inline moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h]:
```suggestion
    * @param error_code an error code that encodes the reason for failure or success
    * @param options An option struct which contains the type of redundancy discretization used. This default
    *                implementation only supports the KinematicSearches::NO_DISCRETIZATION method; requesting any
    *                other will result in failure.
```

File: moveit_kinematics/kdl_kinematics_plugin/include/moveit/kdl_kinematics_plugin/chainiksolver_vel_mimic_svd.hpp [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+   * @param chain the chain to calculate the inverse velocity kinematics for
+   * @param mimic_joints A vector of indices that map each (and every) joint onto the corresponding joint in a
+   * reduced set of joints that do not include the mimic joints. This vector must be of size chain.getNrOfJoints().
+   * E.g. if an arm has 7 joints: j0 to j6. Say j2 mimics (follows) j0. Then, mimic_joints should be: [0 1 0 3 4 5 6]
+   * @param num_mimic_joints The number of joints that are setup to follow other joints
+   * @param position_ik false if you want to solve for the full 6 dof end-effector pose,
+   *        true if you want to solve only for the 3 dof end-effector position.
+   * @param threshold if a singular value is below this value, its inverse is set to zero, default: 0.00001
+   */
+  explicit ChainIkSolverVelMimicSVD(const Chain& chain_,
```
Comment (davetcoleman) [inline moveit_kinematics/kdl_kinematics_plugin/include/moveit/kdl_kinematics_plugin/chainiksolver_vel_mimic_svd.hpp]:
```suggestion
  explicit ChainIkSolverVelMimicSVD(const Chain& chain,
```

File: moveit_kinematics/kdl_kinematics_plugin/include/moveit/kdl_kinematics_plugin/kdl_kinematics_plugin.h [hunk 9]
```diff
@@ -122,84 +120,62 @@ class KDLKinematicsPlugin : public kinematics::KinematicsBase
...
-
-  KDL::Chain kdl_chain_;
+  bool initialized_;  //< Internal variable that indicates whether solver is configured and ready
 
-  unsigned int dimension_; /** Dimension of the group */
-
-  KDL::JntArray joint_min_, joint_max_; /** Joint limits */
-
-  mutable random_numbers::RandomNumberGenerator random_number_generator_;
+  unsigned int dimension_;                        //< Dimension of the group
```
Comment (davetcoleman) [inline moveit_kinematics/kdl_kinematics_plugin/include/moveit/kdl_kinematics_plugin/kdl_kinematics_plugin.h]: nit: why the "<" for comments? seems non-standard

File: moveit_kinematics/kdl_kinematics_plugin/include/moveit/kdl_kinematics_plugin/kdl_kinematics_plugin.h [hunk 10]
```diff
@@ -122,84 +120,62 @@ class KDLKinematicsPlugin : public kinematics::KinematicsBase
...
+  moveit_msgs::KinematicSolverInfo solver_info_;  //< Stores information for the inverse kinematics solver
 
+  const robot_model::JointModelGroup* joint_model_group_;
   robot_state::RobotStatePtr state_;
+  KDL::Chain kdl_chain_;
+  std::unique_ptr<KDL::ChainFkSolverPos> fk_solver;
```
Comment (davetcoleman) [inline moveit_kinematics/kdl_kinematics_plugin/include/moveit/kdl_kinematics_plugin/kdl_kinematics_plugin.h]:
```suggestion
  std::unique_ptr<KDL::ChainFkSolverPos> fk_solver_;
```

File: moveit_kinematics/kdl_kinematics_plugin/src/chainiksolver_vel_mimic_svd.cpp [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+
+// Modified to account for "mimic" joints, i.e. joints whose motion has a
+// linear relationship to that of another joint.
+// Copyright  (C)  2013  Sachin Chitta, Willow Garage
```
Comment (davetcoleman) [inline moveit_kinematics/kdl_kinematics_plugin/src/chainiksolver_vel_mimic_svd.cpp]: Is this actually Sachin's work or a new file modified from Ruben?

File: moveit_kinematics/kdl_kinematics_plugin/src/chainiksolver_vel_mimic_svd.cpp [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+
+#include <moveit/kdl_kinematics_plugin/chainiksolver_vel_mimic_svd.hpp>
+#include <ros/console.h>
+
+namespace
```
Comment (davetcoleman) [inline moveit_kinematics/kdl_kinematics_plugin/src/chainiksolver_vel_mimic_svd.cpp]: anonymous namespace, love it!

File: moveit_kinematics/kdl_kinematics_plugin/src/chainiksolver_vel_mimic_svd.cpp [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+namespace KDL
+{
+ChainIkSolverVelMimicSVD::ChainIkSolverVelMimicSVD(const Chain& chain,
+                                                   const std::vector<kdl_kinematics_plugin::JointMimic>& mimic_joints,
+                                                   bool position_ik, double threshold)
+  : mimic_joints_(mimic_joints)
+  , num_mimic_joints_(countMimicJoints(mimic_joints))
+  , chain_(chain)
+  , jnt2jac_(chain)
+  , svd_(position_ik ? 3 : 6, chain_.getNrOfJoints() - num_mimic_joints_, Eigen::ComputeThinU | Eigen::ComputeThinV)
```
Comment (davetcoleman) [inline moveit_kinematics/kdl_kinematics_plugin/src/chainiksolver_vel_mimic_svd.cpp]: please add comment describing this logic... its very difficult to understand having it nestled in the initalizer list. alternatively, providing a anon helper function to call within the initializer list would be better

File: moveit_kinematics/kdl_kinematics_plugin/include/moveit/kdl_kinematics_plugin/kdl_kinematics_plugin.h (hunk 9 above)
Comment (rhaschke) [reply moveit_kinematics/kdl_kinematics_plugin/include/moveit/kdl_kinematics_plugin/kdl_kinematics_plugin.h]: http://www.doxygen.nl/manual/docblocks.html#memberdoc
But, indeed I missed the third slash...

File: moveit_kinematics/kdl_kinematics_plugin/src/chainiksolver_vel_mimic_svd.cpp (hunk 11 above)
Comment (rhaschke) [reply moveit_kinematics/kdl_kinematics_plugin/src/chainiksolver_vel_mimic_svd.cpp]: I don't know. KDL doesn't provide support for mimic joints. I guess Sachin adapted the standard KDL solver to compute a Jacobian for mimic joints.

File: moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h (hunk 4 above)
Comment (rhaschke) [reply moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h]: I don't like to advertize the redundancy parameters anymore. They are not used by MoveIt.
Rather, they are parameters of IKfast only.
However, I added a generic comment as follows:
`@param options container for other IK options. See definition of KinematicsQueryOptions for details.`

File: moveit_kinematics/kdl_kinematics_plugin/src/chainiksolver_vel_mimic_svd.cpp (hunk 13 above)
Comment (rhaschke) [reply moveit_kinematics/kdl_kinematics_plugin/src/chainiksolver_vel_mimic_svd.cpp]: Done.

File: moveit_kinematics/kdl_kinematics_plugin/include/moveit/kdl_kinematics_plugin/kdl_kinematics_plugin.h (hunk 10 above)
Comment (rhaschke) [reply moveit_kinematics/kdl_kinematics_plugin/include/moveit/kdl_kinematics_plugin/kdl_kinematics_plugin.h]: Yep.

**Comment by rhaschke** [comment]
Merged after having Mike's approval and having Dave's comments addressed.

**Pull request merged**